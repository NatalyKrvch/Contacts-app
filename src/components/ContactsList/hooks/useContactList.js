import { useState, useEffect } from 'react'
import { useGetContactsQuery } from 'services/api/contactsApi'

const useContactList = () => {
  const { data, error, isLoading } = useGetContactsQuery()
  const [toast, setToast] = useState({ show: false, message: '' })

  useEffect(() => {
    if (data && !Array.isArray(data.resources)) {
      setToast({
        show: true,
        message: 'Failed to load contacts. Please try again.',
      })
    }
  }, [data])

  useEffect(() => {
    if (error) {
      setToast({
        show: true,
        message: 'An error occurred while fetching contacts.',
      })
    }
  }, [error])

  const contacts = data?.resources

  return {
    contacts,
    toast,
    isLoading,
    setToast,
  }
}

export default useContactList
