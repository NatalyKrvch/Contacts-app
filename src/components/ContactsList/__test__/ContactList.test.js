import { render, screen, fireEvent } from '@testing-library/react'
import ContactList from '../ContactsList'
import useContactList from '../hooks/useContactList'

jest.mock('../hooks/useContactList')
jest.mock(
  'components/CustomToast/CustomToast',
  () =>
    ({ show, label, message, onClose, bg }) =>
      (
        <div data-testid="custom-toast">
          {label && <strong>{label}</strong>}
          {message && <div>{message}</div>}
          {show && <button onClick={onClose}>Close</button>}
        </div>
      )
)
jest.mock('components/ContactCard/ContactCard', () => ({ contact }) => (
  <div data-testid="contact-card">{contact.name}</div>
))

describe('ContactList', () => {
  const mockSetToast = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders loading spinner when loading', () => {
    useContactList.mockReturnValue({
      contacts: [],
      toast: { show: false, message: '', bg: '', label: '' },
      isLoading: true,
      setToast: mockSetToast,
    })

    const { container } = render(<ContactList />)

    expect(screen.queryByTestId('custom-toast')).not.toBeInTheDocument()
    expect(screen.queryByTestId('contact-card')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders custom toast when no contacts available', () => {
    useContactList.mockReturnValue({
      contacts: null,
      toast: {
        show: true,
        message: 'No contacts available. Please add some',
        bg: '',
        label: 'Oooops',
      },
      isLoading: false,
      setToast: mockSetToast,
    })

    const { container } = render(<ContactList />)

    expect(screen.getByText('Oooops')).toBeInTheDocument()
    expect(
      screen.getByText('No contacts available. Please add some')
    ).toBeInTheDocument()
    expect(screen.getByText('Close')).toBeInTheDocument()

    fireEvent.click(screen.getByText('Close'))

    expect(mockSetToast).toHaveBeenCalledWith({
      show: false,
      message: '',
      bg: '',
      label: '',
    })
    expect(container).toMatchSnapshot()
  })

  it('renders contacts correctly', () => {
    const mockContacts = [
      { id: 1, name: 'John Doe' },
      { id: 2, name: 'Jane Smith' },
    ]
    useContactList.mockReturnValue({
      contacts: mockContacts,
      toast: { show: false, message: '', bg: '', label: '' },
      isLoading: false,
      setToast: mockSetToast,
    })

    const { container } = render(<ContactList />)

    expect(screen.getByText('Contacts')).toBeInTheDocument()
    mockContacts.forEach((contact) => {
      expect(screen.getByText(contact.name)).toBeInTheDocument()
    })
    expect(container).toMatchSnapshot()
  })

  it('renders error toast when there is an error', () => {
    useContactList.mockReturnValue({
      contacts: [],
      toast: {
        show: true,
        message: 'An error occurred',
        bg: '',
        label: 'Error happened',
      },
      isLoading: false,
      setToast: mockSetToast,
    })

    const { container } = render(<ContactList />)

    expect(screen.getByText('Error happened')).toBeInTheDocument()
    expect(screen.getByText('An error occurred')).toBeInTheDocument()
    expect(screen.getByText('Close')).toBeInTheDocument()
    fireEvent.click(screen.getByText('Close'))
    expect(mockSetToast).toHaveBeenCalledWith({
      show: false,
      message: '',
      bg: '',
      label: '',
    })
    expect(container).toMatchSnapshot()
  })
})
