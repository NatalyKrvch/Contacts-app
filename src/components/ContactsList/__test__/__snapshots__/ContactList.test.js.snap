// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList renders contacts correctly 1`] = `
<div>
  <div
    class="container"
    style="max-width: 558px; margin: 0px auto;"
  >
    <p
      class="fw-medium fs-5 mt-3"
    >
      Contacts
    </p>
    <div
      class="mb-3"
    >
      <div
        data-testid="contact-card"
      >
        John Doe
      </div>
    </div>
    <div
      class="mb-3"
    >
      <div
        data-testid="contact-card"
      >
        Jane Smith
      </div>
    </div>
    <div
      data-testid="custom-toast"
    >
      <strong>
        Error happened
      </strong>
    </div>
  </div>
</div>
`;

exports[`ContactList renders custom toast when no contacts available 1`] = `
<div>
  <div
    data-testid="custom-toast"
  >
    <strong>
      Oooops
    </strong>
    <div>
      No contacts available. Please add some
    </div>
    <button>
      Close
    </button>
  </div>
</div>
`;

exports[`ContactList renders error toast when there is an error 1`] = `
<div>
  <div
    class="container"
    style="max-width: 558px; margin: 0px auto;"
  >
    <p
      class="fw-medium fs-5 mt-3"
    >
      Contacts
    </p>
    <div
      data-testid="custom-toast"
    >
      <strong>
        Error happened
      </strong>
      <div>
        An error occurred
      </div>
      <button>
        Close
      </button>
    </div>
  </div>
</div>
`;

exports[`ContactList renders loading spinner when loading 1`] = `
<div>
  <div
    class="spinner-border"
    style="margin: 50px auto;"
  />
</div>
`;
